### PR TITLE
Merge comments while generating metadata from assembly

### PR DESCRIPTION
**Pull request by qinezh**

### CODE REVIEW

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataWorker.cs [hunk 1]
```diff
@@ -441,6 +442,13 @@ private static ExtractMetadataInputModel ValidateInput(ExtractMetadataInputModel
                     var assemblyMetadataValues = from assembly in referencedAssemblyList
                                                  let metadata = GetAssemblyMetadataFromCacheAsync(assemblyFiles, assemblyCompilation, assembly, outputFolder, forceRebuild, _filterConfigFile, assemblyExtension)
                                                  select metadata.Result.Item1;
+                    var commentFiles = (from file in assemblyFiles
+                                        let xmlFile = Path.ChangeExtension(file, SupportedCommentFileExtension)
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataWorker.cs]: `select Path.ChangeExtension(file, SupportedCommentFileExtension) into xmlFile`

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/MergeComments/MergeCommentsHelper.cs [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+    public static class MergeCommentsHelper
+    {
+        public static void MergeComments(IEnumerable<MetadataItem> items, IEnumerable<string> commentFiles)
+        {
+            foreach (var item in items)
+            {
+                var list = from file in commentFiles
+                           let name = Path.GetFileNameWithoutExtension(file)
+                           where name != null && name.Equals(item.Name, StringComparison.OrdinalIgnoreCase)
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/MergeComments/MergeCommentsHelper.cs]: use `string.Equals`

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/MergeComments/UidAndComment.cs [hunk 3]
```diff
@@ -0,0 +1,11 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DocAsCode.Metadata.ManagedReference
+{
+    internal sealed class UidAndComment
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/MergeComments/UidAndComment.cs]: naming

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/MergeComments/UidAndReader.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DocAsCode.Metadata.ManagedReference
+{
+    using System.Xml;
+
+    internal sealed class UidAndReader
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/MergeComments/UidAndReader.cs]: naming

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/MergeComments/UidAndComment.cs (hunk 3 above)
Comment (qinezh) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/MergeComments/UidAndComment.cs]: I copied it from https://github.com/dotnet/docfx/blob/dev/tools/MergeDeveloperComments/Program.cs#L469 ...

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/MergeComments/TripleSlashCommentParserContextNew.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DocAsCode.Metadata.ManagedReference
+{
+    using System;
+
+    using Microsoft.DocAsCode.DataContracts.Common;
+
+    internal sealed class TripleSlashCommentParserContextNew : ITripleSlashCommentParserContext
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/MergeComments/TripleSlashCommentParserContextNew.cs]: >TripleSlashCommentParserContextNew [](start = 26, length = 34)

naming

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/MergeComments/TripleSlashCommentParserContextNew.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    {
+        public static readonly TripleSlashCommentParserContext Instance = new TripleSlashCommentParserContext
+        {
+            AddReferenceDelegate = (s, e) => { },
+            Normalize = true,
+        };
+
+        public Action<string, string> AddReferenceDelegate { get; set; }
+        public bool Normalize { get; set; }
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/MergeComments/TripleSlashCommentParserContextNew.cs]: >Normalize [](start = 20, length = 9)

Where is this used?

Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/MergeComments/TripleSlashCommentParserContextNew.cs]: remove?

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/MergeComments/CommentidAndReader.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace Microsoft.DocAsCode.Metadata.ManagedReference
+{
+    using System.Xml;
+
+    internal sealed class CommentIdAndReader
+    {
+        public string CommentId { get; set; }
+        public XmlReader Reader { get; set; }
+        public CommentIdAndComment ToUidAndElement() => new CommentIdAndComment { CommentId = CommentId, Comment = Reader.ReadOuterXml() };
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/MergeComments/CommentidAndReader.cs]: naming

**Comment by superyyrrzz** [comment]
>             public bool Normalize { get; set; }

Can remove?

---
Refers to: tools/MergeDeveloperComments/Program.cs:490 in d46d692. [](commit_id = d46d692832a360e9f37002e0987a15278dc76d5d, deletion_comment = False)

**Comment by superyyrrzz** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/MergeComments/CommentidAndReader.cs [hunk 8]
```diff
@@ -0,0 +1,14 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DocAsCode.Metadata.ManagedReference
+{
+    using System.Xml;
+
+    internal sealed class CommentIdAndReader
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/MergeComments/CommentidAndReader.cs]: >CommentIdAndReader [](start = 26, length = 18)

Minor: file name "Comment'I'dAndReader"

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/MergeComments/CommentIdAndComment.cs [hunk 9]
```diff
@@ -0,0 +1,11 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DocAsCode.Metadata.ManagedReference
+{
+    internal sealed class CommentIdAndComment
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/MergeComments/CommentIdAndComment.cs]: >CommentIdAndComment [](start = 26, length = 19)

How about naming with "Comment" and have "Id" and "Content"?

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/MergeComments/CommentidAndReader.cs [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+    {
+        public string CommentId { get; set; }
+        public XmlReader Reader { get; set; }
+        public CommentIdAndComment ToCommentIdAndComment() => new CommentIdAndComment { CommentId = CommentId, Comment = Reader.ReadOuterXml() };
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/MergeComments/CommentidAndReader.cs]: >Reader [](start = 121, length = 6)

before ReadOuterXml, check whether Reader is null?

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/MergeComments/CommentIdAndComment.cs (hunk 9 above)
Comment (qinezh) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/MergeComments/CommentIdAndComment.cs]: I think `Comment` is OK here, just like `雷锋塔和雷锋` would not make us confused...

**Comment by vwxyzh** [comment]
:shipit:

**Comment by hellosnow** [comment]
:shipit:

**Pull request merged**